### PR TITLE
Add Android device config validation

### LINKED ISSUE #187: Kotlin | Validate Android versions/models
See https://github.com/TestArmada/flank/blob/master/cloud_testing/kotlin_poc/src/main/kotlin/ftl/ios/IosCatalog.kt

Requires https://github.com/TestArmada/flank/issues/133 to be done first

Flank should error if a request Android device isn't in the catalog.

### PR DESCRIPTION
**Pull request by StephenBrough**
Fixes #187

### CODE REVIEW

File: cloud_testing/kotlin_poc/src/main/kotlin/ftl/android/AndroidCatalog.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    val androidVersionIds by lazy { androidDeviceCatalog.versions.map { it.id } }
+
+
+    fun supportedDeviceConfig(modelId: String, versionId: String): DeviceConfigCheck {
+        if (!androidModelIds.contains(modelId)) return UnsupportedModelId
+        if (!androidVersionIds.contains(versionId)) return UnsupportedVersionId
+
+        val supportedVersionIds = androidDeviceCatalog.models.find { it.id == modelId }?.supportedVersionIds
+        supportedVersionIds?.let {
+            if (!it.contains(versionId)) return IncompatibleModelVersion
```
Comment (Macarse) [inline cloud_testing/kotlin_poc/src/main/kotlin/ftl/android/AndroidCatalog.kt]: shouldn't this just throw an exception?

File: cloud_testing/kotlin_poc/src/main/kotlin/ftl/android/AndroidCatalog.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * Contains lists of possible Android device and version ids, as well as checks
+ * to validate device configuration
+ */
+object AndroidCatalog {
+    private val androidDeviceCatalog by lazy {
+        GcTesting.get.testEnvironmentCatalog().get("android").execute().androidDeviceCatalog
+    }
+
+    val androidModelIds by lazy { androidDeviceCatalog.models.map { it.id } }
+    val androidVersionIds by lazy { androidDeviceCatalog.versions.map { it.id } }
```
Comment (bootstraponline) [inline cloud_testing/kotlin_poc/src/main/kotlin/ftl/android/AndroidCatalog.kt]: do we need?

- androidModelIds
- androidVersionIds

it seems like ` androidDeviceCatalog.models.find ` covers the validation?

File: cloud_testing/kotlin_poc/src/test/kotlin/ftl/android/AndroidCatalogTest.kt [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import org.junit.Test
+import org.junit.rules.ExpectedException
+
+class AndroidCatalogTest {
+
+    private val bitrise = System.getenv("BITRISE_IO") != null
+
+    @Test
+    fun validateDeviceConfig() {
+        if (bitrise) return
```
Comment (Macarse) [inline cloud_testing/kotlin_poc/src/test/kotlin/ftl/android/AndroidCatalogTest.kt]: is this necessary? I haven't seen it in other tests.

File: cloud_testing/kotlin_poc/src/main/kotlin/ftl/android/AndroidCatalog.kt [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+            if (!it.contains(versionId)) return IncompatibleModelVersion
+        } ?: return UnsupportedModelId
+
+        return SupportedDeviceConfig
+    }
+}
+
+sealed class DeviceConfigCheck
+object SupportedDeviceConfig: DeviceConfigCheck()
+object UnsupportedModelId : DeviceConfigCheck()
```
Comment (bootstraponline) [inline cloud_testing/kotlin_poc/src/main/kotlin/ftl/android/AndroidCatalog.kt]: I love the exception classes!

File: cloud_testing/kotlin_poc/src/test/kotlin/ftl/android/AndroidCatalogTest.kt (hunk 3 above)
Comment (bootstraponline) [inline cloud_testing/kotlin_poc/src/test/kotlin/ftl/android/AndroidCatalogTest.kt]: The Android catalog isn't API restricted so this test should work on Bitrise.

File: cloud_testing/kotlin_poc/src/main/kotlin/ftl/android/AndroidCatalog.kt (hunk 2 above)
Comment (bootstraponline) [reply cloud_testing/kotlin_poc/src/main/kotlin/ftl/android/AndroidCatalog.kt]: nevermind, it looks like the error messages are a bit cleaner with this.

File: cloud_testing/kotlin_poc/src/test/kotlin/ftl/android/AndroidCatalogTest.kt (hunk 3 above)
Comment (StephenBrough) [reply cloud_testing/kotlin_poc/src/test/kotlin/ftl/android/AndroidCatalogTest.kt]: cool, I'll get rid of it then!

Comment (StephenBrough) [reply cloud_testing/kotlin_poc/src/test/kotlin/ftl/android/AndroidCatalogTest.kt]: nope, I'll get rid of it :)

File: cloud_testing/kotlin_poc/src/main/kotlin/ftl/android/AndroidCatalog.kt (hunk 1 above)
Comment (StephenBrough) [reply cloud_testing/kotlin_poc/src/main/kotlin/ftl/android/AndroidCatalog.kt]: exceptions are thrown when the run command is processed

**Comment by bootstraponline** [comment]
> ftl.android.AndroidCatalogTest > validateDeviceConfig FAILED

Sorry, looks like we do need that bitrise check in the test after all. The job can't have secrets since we're running on Pull Requests.

File: cloud_testing/kotlin_poc/src/test/kotlin/ftl/android/AndroidCatalogTest.kt (hunk 3 above)
Comment (StephenBrough) [reply cloud_testing/kotlin_poc/src/test/kotlin/ftl/android/AndroidCatalogTest.kt]: ...it's back!

Comment (StephenBrough) [reply cloud_testing/kotlin_poc/src/test/kotlin/ftl/android/AndroidCatalogTest.kt]: looks like we need it, per @bootstraponline's comment below

**Pull request merged**